finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala

```diff
@@ -306,208 +211,170 @@ private[h2] trait Netty4StreamTransport[SendMsg <: Message, RecvMsg <: Message]
-
-          case state@LocalClosed(remote@RemotePending()) =>
-            val msg = mkRecvMsg(hdrs.headers, NilStream)
+          // Pending
+          case StreamState(sendState, RecvPending) =>
+            val stream = Stream.empty()
+            val msg = mkRecvMsg(hdrs.headers, stream)
             if (ConnectionHeaders.detect(msg.headers)) {
-              if (resetFromLocal(state, Reset.ProtocolError)) true
-              else recv(hdrs)
+              log.debug("[%s] illegal connection headers detected in %s", prefix, msg.headers)
+              reset(Reset.ProtocolError, local = true)
+              false
             } else {
-              if (stateRef.compareAndSet(state, Closed(Reset.NoError))) {
-                remote.setMessage(msg)
-                resetP.setDone()
+              if (stateRef.compareAndSet(state, StreamState(sendState, RecvClosed))) {
+                stream.onCancel.onSuccess { rst =>
+                  // If the recvMsg stream is cancelled, reset the stream.
+                  reset(rst, local = true); ()
+                }
+                recvMsg.setValue(msg)
+                if (sendState == SendClosed) {
+                  resetP.setDone()
+                }
                 true
               } else recv(hdrs)
             }
 
-          case LocalClosed(remote@RemoteStreaming()) =>
-            if (stateRef.compareAndSet(state, Closed(Reset.NoError))) {
-              if (recvFrame(toFrame(hdrs), remote)) {
-                remote.close()
+          // Streaming
+          case StreamState(sendState, RecvStreaming(q)) =>
+            if (stateRef.compareAndSet(state, StreamState(sendState, RecvClosed))) {
+              if (sendState == SendClosed) {
                 resetP.setDone()
-                true
-              } else false
+              }
+              recvFrame(toFrame(hdrs), q)
             } else recv(hdrs)
+
+          // Closed
+          case StreamState(sendState, RecvClosed) =>
+            log.debug("[%s] frame received while in RecvClosed", prefix)
+            reset(Reset.InternalError, local = true)
+            false
         }
 
+      /*
+       * HEADERS FRAME
+       */
       case hdrs: Http2HeadersFrame =>
         // A HEADERS frame without END_STREAM may only be received to
         // initiate a message (i.e. when the remote is still pending).
         state match {
-          case Closed(_) => false
-
-          case state@RemoteClosed() =>
-            if (resetFromLocal(state, Reset.Closed)) false
-            else recv(hdrs)
-
-          case RemoteOpen(remote@RemoteStreaming()) =>
-            if (resetFromLocal(remote, Reset.InternalError)) false
-            else recv(hdrs)
-
-          case Open(remote@RemotePending()) =>
+          // Pending
+          case StreamState(sendState, RecvPending) =>
             val q = new AsyncQueue[Frame]
-            val msg = mkRecvMsg(hdrs.headers, Stream(q))
+            val stream = Stream(q)
```

Comment:
here (and throughout), might be helpful to call this `recvStream` -- it can be a bit confusing which stream we're talking about at times.